### PR TITLE
feat: Added data-safe stack trace collection to telemetry

### PR DESCRIPTION
**Pull request by splouisliu**
#### Which issue(s) does this change fix?
N.A.

#### Why is this change necessary?
Previously, only exit code and exit reason are collected when an exception occurs during command executions. In order for us to better understand errors that occur on the user's end, we should also collect stack traces.

#### How does it address the issue?
Stack trace is now collected when an exception occurs during command executions. The paths contained within are also cleaned to remove user-sensitive information. The final stack trace and exception message are attached to telemetry.

#### Mandatory Checklist
**PRs will only be reviewed after checklist is complete**

- [x] Add input/output [type hints](https://docs.python.org/3/library/typing.html) to new functions/methods
- [ ] Write design document if needed ([Do I need to write a design document?](https://github.com/aws/aws-sam-cli/blob/develop/DEVELOPMENT_GUIDE.md#design-document))
- [x] Write/update unit tests
- [x] Write/update integration tests
- [ ] Write/update functional tests if needed
- [x] `make pr` passes
- [ ] `make update-reproducible-reqs` if dependencies were changed
- [ ] Write documentation

By submitting this pull request, I confirm that my contribution is made under the terms of the [Apache 2.0 license](https://www.apache.org/licenses/LICENSE-2.0).

### CODE REVIEW

**Review by hawflau — CHANGES_REQUESTED** [review]

File: samcli/lib/telemetry/metric.py [hunk 1]
```diff
@@ -170,6 +176,44 @@ def wrapped(*args, **kwargs):
...
+        # If working directory is found within path, remove it
+        dir_index = path.find(working_directory)
+        if dir_index != -1:
+            stack_trace_lines[
+                i
+            ] = f'{prefix}"{path[:dir_index]}{path[dir_index + len(working_directory) + 1 :]}"{suffix}'
+            continue
+
+        # If working directory is not found within path, replace leading path of the last file with: /../ or ..\
+        # i.e. /Users/user1/test.py becomes /../test.py
```
[the remark this answers is not in the document]
Comment (splouisliu) [reply samcli/lib/telemetry/metric.py:207]: The problem is that sometimes the path leads outside the aws-sam-cli folder (i.e. to venv folder). It wouldn't be possible to do this unless we know how the user's outside file directories are structured. AFAIK, Python doesn't have a functionality to return relative path in the stack trace itself.

[the remark this answers is not in the document]
Comment (hawflau) [reply samcli/lib/telemetry/metric.py:207]: I think there's a way to capture more information than keeping only the filename
For example, 
1. Tracing `samcli` - e.g. `/Volumes/workplace/gh/aws-sam-cli/samcli/lib/deploy/deployer.py` -> `/../samcli/lib/deploy/deployer.py`
2. Tracing anything after `site-packages` - e.g. `/usr/local/Cellar/aws-sam-cli/1.49.0/libexec/lib/python3.8/site-packages/botocore/waiter.py` -> `/../botocore/waiter.py`
3. For other cases, it makes sense to just keep the filename as you've implemented.

Handling the cases 1 and 2 should improve the debugging experience a lot.

File: samcli/lib/telemetry/metric.py [hunk 2]
```diff
@@ -170,6 +175,48 @@ def wrapped(*args, **kwargs):
...
+    Returns
+    -------
+    str
+        Stack trace with paths filtered
+    """
+    stack_trace_lines = traceback.format_exc().splitlines()
+
+    for i, line in enumerate(stack_trace_lines):
+        # Make no changes if current line does not contain a path
+        line_split = line.split('"')
```
Comment (hawflau) [inline samcli/lib/telemetry/metric.py:191]: Will this example break the path detection logic?
```
Traceback (most recent call last):
  File "test.py", line 3, in <module>
    print(a["b"])
```

**Review by hawflau — CHANGES_REQUESTED** [review]
LGTM overall, just want to clarify about one more potential edge case.

File: tests/unit/lib/telemetry/test_metric.py [hunk 3]
```diff
@@ -227,8 +230,11 @@ def real_fn():
             "Measured duration must be in milliseconds and greater than equal to the sleep duration",
         )
 
+    @patch("samcli.lib.telemetry.metric._get_stack_trace")
     @patch("samcli.lib.telemetry.metric.Context")
-    def test_must_record_user_exception(self, ContextMock):
+    def test_must_record_user_exception(self, ContextMock, StackTraceMock):
```
Comment (hawflau) [inline tests/unit/lib/telemetry/test_metric.py:235]: Suggest to match name casing with the method/class being patched (i.e. method -> snake_case, class -> CamelCase)

```suggestion
    def test_must_record_user_exception(self, ContextMock, get_stack_trace_mock):
```

File: tests/unit/lib/telemetry/test_metric.py [hunk 4]
```diff
@@ -244,14 +250,20 @@ def real_fn():
...
+        )
         args, _ = self.telemetry_instance.emit.call_args_list[0]
         metric = args[0]
         assert metric.get_metric_name() == "commandRun"
         self.assertGreaterEqual(metric.get_data().items(), expected_attrs.items())
 
+    @patch("samcli.lib.telemetry.metric._get_stack_trace")
     @patch("samcli.lib.telemetry.metric.Context")
-    def test_must_record_wrapped_user_exception(self, ContextMock):
+    def test_must_record_wrapped_user_exception(self, ContextMock, StackTraceMock):
```
Comment (hawflau) [inline tests/unit/lib/telemetry/test_metric.py:264]:
```suggestion
    def test_must_record_wrapped_user_exception(self, ContextMock, get_stack_trace_mock):
```

**Review by hawflau — CHANGES_REQUESTED** [review]
LGTM overall. Just some minor suggestions on naming. Should be good after fixing them.

**Review by moelasmar — APPROVED** [review]

File: samcli/lib/telemetry/metric.py [hunk 5]
```diff
@@ -170,6 +175,58 @@ def wrapped(*args, **kwargs):
...
+    str
+        Stack trace with paths cleaned
+    """
+    tb_exception = traceback.TracebackException.from_exception(exception)
+    _clean_stack_summary_paths(tb_exception.stack)
+    stack_trace = "".join(list(tb_exception.format()))
+    return stack_trace
+
+
+def _clean_stack_summary_paths(stack_summary: traceback.StackSummary):
```
Comment (qingchm) [inline samcli/lib/telemetry/metric.py:198]: nit: return type -> None

**Review by qingchm — APPROVED** [review]

File: samcli/lib/telemetry/metric.py [hunk 6]
```diff
@@ -170,6 +177,62 @@ def wrapped(*args, **kwargs):
...
+
+    Returns
+    -------
+    (str, str)
+        (stack trace, exception message)
+    """
+    tb_exception = traceback.TracebackException.from_exception(exception)
+    _clean_stack_summary_paths(tb_exception.stack)
+    stack_trace = "".join(list(tb_exception.format()))
+    exception_msg = list(tb_exception.format_exception_only())[-1]
```
Comment (hawflau) [inline samcli/lib/telemetry/metric.py:199]: Are we sure the list always has at least 1 element?

Comment (splouisliu) [reply samcli/lib/telemetry/metric.py:199]: Yup: https://docs.python.org/3/library/traceback.html#traceback.format_exception_only
